src/levanter/models/llama.py

```diff
@@ -0,0 +1,475 @@
+        self.Pos = Pos
+        self.cos_cached, self.sin_cached = self._get_cos_sin_cache(Pos=Pos, HeadSize=HeadSize, base=base)
+
+    @staticmethod
+    def _get_cos_sin_cache(HeadSize: NamedArray, Pos: NamedArray, base: float) -> Tuple[jnp.ndarray, jnp.ndarray]:
+        HeadHalfSize = HeadSize.resize(HeadSize.size // 2)
+        inv_freq: NamedArray = 1.0 / (base ** (hax.arange(HeadHalfSize, step=2) / HeadSize.size))
+
+        position_ids: NamedArray = hax.arange(Pos)
+
+        # Evaluates the Einstein summation convention on the operands.
+        freqs = position_ids * inv_freq.broadcast_axis(Pos)
+        # This is different from the paper but alignes with HF implementation:
+        # It uses a different permutation in order to obtain the same calculation
+        emb = hax.concatenate(HeadSize, (freqs, freqs))
+        cos_cached = hax.cos(emb)
+        sin_cached = hax.sin(emb)
+        return cos_cached, sin_cached
+
+    def __call__(self, seq_len: int) -> Tuple[NamedArray, NamedArray]:
+        return (
+            self.cos_cached[self.Pos, :seq_len],
+            self.sin_cached[self.Pos, :seq_len],
+        )
+
+
+class LlamaAttention(StateDictSerializationMixin, eqx.Module):
+    config: LlamaConfig = eqx.static_field()
+    q_proj: hnn.Linear  # projection from Embed to query
+    k_proj: hnn.Linear  # projection from Embed to key
+    v_proj: hnn.Linear  # projection from Embed to value
+    o_proj: hnn.Linear  # projection from Heads to output
+    rotary_emb: LlamaRotaryEmbedding  # rotary embedding
+
+    @staticmethod
+    def init(config: LlamaConfig, *, key) -> "LlamaAttention":
+        use_bias = config.use_bias
+        Embed = config.Embed
+        k_q, k_k, k_v, k_o = jrandom.split(key, 4)
+        q_proj = hnn.Linear.init(In=Embed, Out=(config.Heads, config.HeadSize), key=k_q, use_bias=use_bias)
+        k_proj = hnn.Linear.init(In=Embed, Out=(config.Heads, config.HeadSize), key=k_k, use_bias=use_bias)
+        v_proj = hnn.Linear.init(In=Embed, Out=(config.Heads, config.HeadSize), key=k_v, use_bias=use_bias)
+        o_proj = hnn.Linear.init(In=(config.Heads, config.HeadSize), Out=Embed, key=k_o, use_bias=use_bias)
+        rotary_emb = LlamaRotaryEmbedding(config.HeadSize, config.Pos)
+        return LlamaAttention(config, q_proj, k_proj, v_proj, o_proj, rotary_emb)
+
+    @named_call
+    def __call__(self, x: NamedArray, mask: Optional[NamedArray]):
+        q = self.q_proj(x)  # TODO: rearrange and possibly rename
+        k = self.k_proj(x)
+        v = self.v_proj(x)
+
+        cos, sin = self.rotary_emb(seq_len=self.config.seq_len)
+
+        q, k = _apply_rotary_pos_emb(q, k, cos, sin)
+
+        scale = jax.lax.rsqrt(float(self.config.HeadSize.size))
+
+        # do this first to help keep FP values small
+        q = q * scale
+
+        q = q.astype(jnp.float32)
+        k = k.astype(jnp.float32)
+        k = k.rename({"position": "key_position"})
```

Comment:
you need to rename v's position too

Reply:
i think you can replace line s 240-258 with `haliax.nn.attention.dot_product_attention(config.Pos, config.Pos, config.HeadDim, q, k, v, mask)`